src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -0,0 +1,668 @@
+-- overlays []        == 'empty'
+-- overlays [x]       == x
+-- overlays [x,y]     == 'overlay' x y
+-- overlays           == 'foldr' 'overlay' 'empty'
+-- 'isEmpty' . overlays == 'all' 'isEmpty'
+-- @
+overlays :: (Ord a, Ord b) => [AdjacencyMap a b] -> AdjacencyMap a b
+overlays ams = BAM (Map.unionsWith Set.union (map leftAdjacencyMap ams)) $
+                    Map.unionsWith Set.union (map rightAdjacencyMap ams)
+
+-- | Connect a given list of graphs.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- connects []        == 'empty'
+-- connects [x]       == x
+-- connects [x,y]     == connect x y
+-- connects           == 'foldr' 'connect' 'empty'
+-- 'isEmpty' . connects == 'all' 'isEmpty'
+-- @
+connects :: (Ord a, Ord b) => [AdjacencyMap a b] -> AdjacencyMap a b
+connects = foldr connect empty
+
+-- | Swap parts of a given graph.
+-- Complexity: /O(1)/ time and memory.
+--
+-- @
+-- swap 'empty'        == 'empty'
+-- swap . 'leftVertex' == rightVertex
+-- swap . 'vertices'   == flip 'vertices'
+-- swap ('edge' 1 "a") == 'edge' "a" 1
+-- swap . 'edges'      == 'edges' . map Data.Tuple.'Data.Tuple.swap'
+-- swap . swap       == id
+-- @
+swap :: AdjacencyMap a b -> AdjacencyMap b a
+swap (BAM lr rl) = BAM rl lr
+
+-- | Construct a bipartite 'AdjacencyMap' from "Algebra.Graph.AdjacencyMap"
+-- with given part identifiers, adding all needed edges to make the graph
+-- undirected and removing all edges inside one part.
+-- Complexity: /O(m log(n))/.
+--
+-- @
+-- toBipartite (Algebra.Graph.AdjacencyMap.'Algebra.Graph.AdjacencyMap.empty')                     == 'empty'
+-- toBipartite (Algebra.Graph.AdjacencyMap.'Algebra.Graph.AdjacencyMap.edge' (Left 1) (Right 1))   == 'edge' 1 1
+-- toBipartite (Algebra.Graph.AdjacencyMap.'Algebra.Graph.AdjacencyMap.edge' (Left 1) (Left 1))    == 'empty'
+-- toBipartite (Algebra.Graph.AdjacencyMap.'Algebra.Graph.AdjacencyMap.edge' (Left 1) (Right "a")) == 'edge' 1 "a"
+-- @
+toBipartite :: (Ord a, Ord b) => AM.AdjacencyMap (Either a b) -> AdjacencyMap a b
+toBipartite m = BAM (Map.fromAscList [ (u, setRights vs) | (Left  u, vs) <- Map.toAscList (AM.adjacencyMap $ AM.symmetricClosure m)])
+                         (Map.fromAscList [ (u, setLefts  vs) | (Right u, vs) <- Map.toAscList (AM.adjacencyMap $ AM.symmetricClosure m)])
+    where
+        setRights = Set.fromAscList . rights . Set.toAscList
+        setLefts  = Set.fromAscList . lefts  . Set.toAscList
+
+-- | Construct an 'Algrebra.Graph.AdjacencyMap' from a bipartite
+-- 'AdjacencyMap'.
+-- Complexity: /O(m log(n))/.
+--
+-- @
+-- fromBipartite 'empty'          == 'Algebra.Graph.AdjacencyMap.empty'
+-- fromBipartite ('leftVertex' 1) == 'Algebra.Graph.AdjacencyMap.vertex' (Left 1)
+-- fromBipartite ('edge' 1 2)     == 'Algebra.Graph.AdjacencyMap.edges' [(Left 1, Right 2), (Right 2, Left 1)]
+-- @
+fromBipartite :: (Ord a, Ord b) => AdjacencyMap a b -> AM.AdjacencyMap (Either a b)
+fromBipartite (BAM lr rl) = AM.overlays $
```

Comment:
I think you could use `fromAdjacencySets` for faster implementation:

https://hackage.haskell.org/package/algebraic-graphs-0.4/docs/Algebra-Graph-AdjacencyMap.html#v:fromAdjacencySets